src/sys/fanotify.rs

```diff
@@ -0,0 +1,365 @@
+    pub response: Response,
+}
+
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+/// Response to be wrapped in `FanotifyResponse` and sent to the `Fanotify`
+/// group to allow or deny an event.
+pub enum Response {
+    /// Allow the event.
+    Allow,
+    /// Deny the event.
+    Deny,
+}
+
+/// A fanotify group. This is also a file descriptor that can feed to other
+/// interfaces consuming file descriptors.
+#[derive(Debug)]
+pub struct Fanotify {
+    fd: OwnedFd,
+}
+
+impl Fanotify {
+    /// Initialize a new fanotify group.
+    ///
+    /// Returns a Result containing a Fanotify instance.
+    ///
+    /// For more information, see [fanotify_init(2)](https://man7.org/linux/man-pages/man7/fanotify_init.2.html).
+    pub fn init(flags: InitFlags, event_f_flags: OFlags) -> Result<Fanotify> {
+        let res = Errno::result(unsafe {
+            libc::fanotify_init(flags.bits(), event_f_flags.bits())
+        });
+        res.map(|fd| Fanotify { fd: unsafe { OwnedFd::from_raw_fd(fd) }})
+    }
+
+    /// Add, remove, or modify an fanotify mark on a filesystem object.
+    /// If `dirfd` is `None`, `AT_FDCWD` is used.
+    ///
+    /// Returns a Result containing either `()` on success or errno otherwise.
+    ///
+    /// For more information, see [fanotify_mark(2)](https://man7.org/linux/man-pages/man7/fanotify_mark.2.html).
+    pub fn mark<P: ?Sized + NixPath>(
+        &self,
+        flags: MarkFlags,
+        mask: MaskFlags,
+        dirfd: Option<RawFd>,
+        path: Option<&P>,
+    ) -> Result<()> {
+        fn with_opt_nix_path<P, T, F>(p: Option<&P>, f: F) -> Result<T>
+        where
+            P: ?Sized + NixPath,
+            F: FnOnce(*const libc::c_char) -> T,
+        {
+            match p {
+                Some(path) => path.with_nix_path(|p_str| f(p_str.as_ptr())),
+                None => Ok(f(std::ptr::null())),
+            }
+        }
+
+        let res = with_opt_nix_path(path, |p| unsafe {
+            libc::fanotify_mark(
+                self.fd.as_raw_fd(),
+                flags.bits(),
+                mask.bits(),
+                dirfd.unwrap_or(libc::AT_FDCWD),
+                p,
+            )
+        })?;
+
+        Errno::result(res).map(|_| ())
+    }
+
+    /// Read incoming events from the fanotify group.
+    ///
+    /// Returns a Result containing either a `Vec` of events on success or errno
+    /// otherwise.
+    ///
+    /// # Errors
+    ///
+    /// Possible errors can be those that are explicitly listed in
+    /// [fanotify(2)](https://man7.org/linux/man-pages/man7/fanotify.2.html) in
+    /// addition to the possible errors caused by `read` call.
+    /// In particular, `EAGAIN` is returned when no event is available on a
+    /// group that has been initialized with the flag `InitFlags::FAN_NONBLOCK`,
+    /// thus making this method nonblocking.
+    pub fn read_events(&self) -> Result<Vec<FanotifyEvent>> {
```

Comment:
Per the man page (man 7 fanotify), the buffer may not just contain the `struct fanotify_event_metadata` type, should we check and expose them here

Reply:
Yes, we eventually should. But for now, I'm working on another libc contribution as the needed structs are not declared in it yet. I let you decide if this first version of fanotify which contains features up to Linux 5.1 is ok to be merged and completed in another incoming contribution or if you prefer to wait for libc to be updated to get the whole fanotify package in this PR.

Reply:
This PR looks good, let's merge it first:)